File: loopback/core/redact.js
'use strict';
/*
 * redact (core) — client-side redaction (design §7).
 *
 * Strips PII / secrets from a candidate excerpt BEFORE it is shown in the consent
 * gate and BEFORE it leaves the machine. The output is exactly what the consent
 * gate displays and exactly what is sent (show-exactly-what-is-sent).
 *
 * Extracted verbatim from the former scripts/redact so the behavior (and the
 * fixtures/redact-cases.json contract) is unchanged; now an in-process pure
 * function reused by the MCP server, the CLI, and the submit path.
 *
 * Idempotent: re-running on already-redacted text is a no-op. The [redacted-*]
 * placeholders never re-match any pattern.
 */

const MAX_LEN = 600; // excerpt cap in BYTES (matches the consent-gate budget)

function byteLen(str) {
  return Buffer.byteLength(str, 'utf8');
}

// Slice to at most `maxBytes` bytes without splitting a multi-byte UTF-8 char.
function sliceBytes(str, maxBytes) {
  const buf = Buffer.from(str, 'utf8');
  if (buf.length <= maxBytes) return str;
  let end = maxBytes;
  while (end > 0 && (buf[end] & 0xc0) === 0x80) end--;
  return buf.slice(0, end).toString('utf8');
}

// Order matters: emails and tokens first (most specific), then paths, then files,
// then user/home tokens. Each replacement target is placeholder-safe.
function redactText(input) {
  let s = String(input);

  // 1. Emails.
  s = s.replace(/[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/g, '[redacted-email]');

  // 2. Known secret/token shapes.
  s = s.replace(/\bgh[pousr]_[A-Za-z0-9]{16,}\b/g, '[redacted-token]');
  s = s.replace(/\bgithub_pat_[A-Za-z0-9_]{20,}\b/g, '[redacted-token]');
  s = s.replace(/\bAKIA[0-9A-Z]{12,}\b/g, '[redacted-token]');
  s = s.replace(/\bsk-[A-Za-z0-9_-]{16,}\b/g, '[redacted-token]');
  s = s.replace(/\bxox[baprs]-[A-Za-z0-9-]{10,}\b/g, '[redacted-token]');
  s = s.replace(/\bey[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{6,}\b/g, '[redacted-token]');
  s = s.replace(/\b[Bb]earer\s+[A-Za-z0-9._-]{8,}/g, 'Bearer [redacted-token]');
  s = s.replace(/\b[A-Fa-f0-9]{32,}\b/g, '[redacted-token]');
  s = s.replace(/\b[A-Za-z0-9+/]{40,}={0,2}\b/g, '[redacted-token]');

  // 3. Filesystem paths — only genuinely path-like tokens, never ordinary
  //    `/`-delimited prose (e.g. "Problem/Solution/Metrics", "TCP/IP").
  const PATH_TAIL = "[^\\s\"'`<>|]*[^\\s\"'`<>|.,;:!?)]";
  const BOUND = "(?:^|(?<=[\\s\"'`(<\\[{]))"; // start-of-token boundary (lookbehind)
  s = s.replace(
    /\b[\w.~-]+(?:\/[\w.~-]+)+\.(?:ts|tsx|js|jsx|mjs|cjs|py|rb|go|rs|java|kt|c|cc|cpp|h|hpp|cs|php|swift|md|json|yaml|yml|toml|ini|env|txt|sql|sh|html|css|scss)\b/gi,
    '[redacted-path]'
  );
  s = s.replace(
    new RegExp('\\S*\\/(?:Users|home|var|tmp|etc|opt|private)\\/' + PATH_TAIL, 'g'),
    '[redacted-path]'
  );
  s = s.replace(
    new RegExp(BOUND + '(?:[A-Za-z]:\\\\|\\\\\\\\|~\\/|\\.{1,2}\\/|\\/)' + PATH_TAIL, 'g'),
    '[redacted-path]'
  );

  // 4. Bare relative file names with a code/doc extension (e.g. auth.ts, prd.md).
  s = s.replace(
    /\b[\w-]+\.(?:ts|tsx|js|jsx|mjs|cjs|py|rb|go|rs|java|kt|c|cc|cpp|h|hpp|cs|php|swift|md|json|yaml|yml|toml|ini|env|txt|sql|sh|html|css|scss)\b/gi,
    '[redacted-file]'
  );

  // 5. Usernames / home references.
  s = s.replace(/\$HOME\b/g, '[redacted-user]');
  s = s.replace(/\bhome directory of [\w.-]+/gi, 'home directory of [redacted-user]');
  if (process.env.USER) {
    const u = process.env.USER.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    if (u) s = s.replace(new RegExp('\\b' + u + '\\b', 'g'), '[redacted-user]');
  }
  // Collapse repeated placeholders that landed adjacent.
  s = s.replace(/(\[redacted-path\])(\[redacted-(?:file|user)\])+/g, '$1');

  // 6. Cap length last, on a clean boundary so we never split a placeholder.
  if (byteLen(s) > MAX_LEN) {
    const SUFFIX = ' …';
    const budget = MAX_LEN - byteLen(SUFFIX);
    s = sliceBytes(s, budget);
    const lastOpen = s.lastIndexOf('[redacted');
    const lastClose = s.lastIndexOf(']');
    if (lastOpen > lastClose) s = s.slice(0, lastOpen);
    s = s.replace(/\s+$/, '') + SUFFIX;
  }

  return s;
}

// Same as redactText but preserves undefined/null/'' inputs (the submit path
// must not turn an absent optional field into the string "undefined").
function redactMaybe(text) {
  if (text === undefined || text === null || text === '') return text;
  return redactText(text);
}

module.exports = { redactText, redactMaybe, MAX_LEN };

File: service/app/redact.py
"""Server-side redaction re-check (design §7, defense in depth).

The client redacts before display/send (show-exactly-what-is-sent). The service
re-runs the same class of detectors so a buggy or malicious client cannot leak
PII/secrets into the central store. Records that still contain PII/secret
patterns are QUARANTINED (HTTP 422), never written to the main ``records`` table.

This module deliberately mirrors the client-side pattern set (plugin A5). It is
detection-only here: the contract is that a clean payload passes unchanged; a
dirty payload is rejected, not silently scrubbed (the user already approved the
exact bytes, so the server's job is to catch leaks, not to mutate consented
content).
"""
from __future__ import annotations

import re
from typing import List

# Each pattern is (name, compiled regex). Order is not significant.
#
# Path detection mirrors the client redactor (plugin A5): only GENUINELY path-like
# tokens count, never ordinary "/"-delimited prose ("Problem/Solution/Metrics",
# "TCP/IP", "and/or"). A token is path-like only when it is rooted in a known
# absolute root, anchored at a token boundary (/, ~/, ./, ../, C:\, \\share), or
# is a slashed token whose basename has a code/doc file extension.
_FILE_EXT = (
    r"ts|tsx|js|jsx|mjs|cjs|py|rb|go|rs|java|kt|c|cc|cpp|h|hpp|cs|php|swift|"
    r"md|json|yaml|yml|toml|ini|env|txt|sql|sh|html|css|scss"
)
_PATTERNS = [
    # Emails: user@host.tld
    ("email", re.compile(r"[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}")),
    # Rooted absolute POSIX paths under a known system root.
    (
        "posix_user_path",
        re.compile(r"/(?:Users|home|var|tmp|etc|opt|private)/[^\s\"'`<>|]+"),
    ),
    # Anchored POSIX paths at a token boundary: leading /, ~/, ./, ../
    (
        "posix_anchored_path",
        re.compile(r"(?:^|(?<=[\s\"'`(<\[{]))(?:~/|\.{1,2}/|/)[^\s\"'`<>|]+"),
    ),
    # Slashed token whose basename carries a code/doc extension (a/b/c.py).
    (
        "extension_path",
        re.compile(
            r"\b[\w.~-]+(?:/[\w.~-]+)+\.(?:" + _FILE_EXT + r")\b",
            re.IGNORECASE,
        ),
    ),
    # Windows paths (drive-letter or UNC share).
    (
        "windows_user_path",
        re.compile(r"(?:[A-Za-z]:\\|\\\\)[^\s\"'`<>|]+"),
    ),
    # GitHub personal access tokens
    ("github_token", re.compile(r"gh[pousr]_[A-Za-z0-9]{16,}")),
    # AWS access key ids
    ("aws_key", re.compile(r"AKIA[0-9A-Z]{16}")),
    # OpenAI-style secret keys
    ("openai_key", re.compile(r"sk-[A-Za-z0-9]{16,}")),
    # Slack tokens
    ("slack_token", re.compile(r"xox[baprs]-[A-Za-z0-9\-]{10,}")),
    # Generic bearer / api key assignments
    ("bearer", re.compile(r"(?i)bearer\s+[A-Za-z0-9._\-]{16,}")),
    # Private key blocks
    ("private_key", re.compile(r"-----BEGIN [A-Z ]*PRIVATE KEY-----")),
]


def find_leaks(text: str | None) -> List[str]:
    """Return the names of any PII/secret patterns present in ``text``."""
    if not text:
        return []
    hits: List[str] = []
    for name, pat in _PATTERNS:
        if pat.search(text):
            hits.append(name)
    return hits


def has_leak(*texts: str | None) -> bool:
    """True if any of the given text fields still contains a PII/secret pattern."""
    return any(find_leaks(t) for t in texts)


def leak_reasons(*texts: str | None) -> List[str]:
    """Aggregate, de-duplicated list of leak pattern names across fields."""
    seen: List[str] = []
    for t in texts:
        for name in find_leaks(t):
            if name not in seen:
                seen.append(name)
    return seen
